### PR TITLE
Preserve caller context across (and improve) deferred reads

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
The transaction context was not saved/restored when dealing with
deferred reads initiated by events like the DelayPools::Update() event.
To fix this, we refactored MemObject::delayread() and its descendants to
use an AsyncCall, which automatically stores/restores code context.

Using explicit async callbacks highlighted the danger of passing
Connection object via CommRead that does not maintain a closure
callback. There was also a related "stuck transaction" suspicion
documented in DeferredReadManager::kickARead(). Fortunately, all these
problems could now be solved by removing DeferredRead and CommRead
classes! The delayed readers already store the Connection object,
maintain closure callbacks, and have to check stored Connection validity
before reading anyway. The general/centralized delayed reading logic is
not really about reading and Connections (those parts are handled by
transaction-specific code) but about triggering reading _attempts_.
Asynchronous calls are perfect (and sufficient) for doing that.

Also fixed Delay Pools for Gopher: delayAwareRead() was initiated only
once from gopherSendComplete() and the subsequent read calls were
delay-unaware (i.e. immediate) reads.

Also fixed a Delay Pools problem with active transactions: A transaction
started with Delay Pools on becomes stuck if a reconfiguration turns
Delay Pools off.

### CODE REVIEW

File: src/http.h [hunk 1]
```diff
@@ -144,7 +146,6 @@ class HttpStateData : public Client
     static bool decideIfWeDoRanges (HttpRequest * orig_request);
     bool peerSupportsConnectionPinning() const;
     const char *blockSwitchingProtocols(const HttpReply&) const;
-
```
Comment (rousskov) [inline src/http.h:147]: Please remove this out of scope change.

File: src/tunnel.cc [hunk 2]
```diff
@@ -858,7 +858,7 @@ TunnelStateData::copyRead(Connection &from, IOCB *completion)
     int bw = from.bytesWanted(1, SQUID_TCP_SO_RCVBUF);
     // XXX: Delay pools must not delay client-to-Squid traffic (i.e. when
     // from.readPendingFunc is tunnelDelayedClientRead()).
-    // XXX: Bug #4913: Use DeferredRead instead.
+    // XXX: Bug #4913: for delay pools, use delayDead() API instead.
```
Comment (rousskov) [inline src/tunnel.cc:861]:
```suggestion
    // XXX: Bug #4913: for delay pools, use delayRead() API instead.
```

**Review by rousskov — CHANGES_REQUESTED** [review]
> PR description: This refactoring revealed that there are too many Connection closure checks in DeferredReadManager and call recipients

IIRC, it was not so much the excessive number of closure checks that triggered further refactoring, but the fact that some Connection objects were stored without a closing callback and, hence, could become stale, making any further closure checks (using that stale object) unsafe. We wanted to remove Connection storage to address suspicions that such problems exist, especially since we were adding additional asynchronous calls. No need to rephrase the current PR description, but please let me know whether my recollection matches yours.

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (af3b6e8).

File: src/clients/Client.h [hunk 3]
```diff
@@ -155,6 +155,10 @@ class Client:
     size_t calcBufferSpaceToReserve(const size_t space, const size_t wantSpace) const;
 
     void adjustBodyBytesRead(const int64_t delta);
+    /// A callback used by delayRead(); implement in sublclasses calling delayRead().
+    virtual void noteDelayedRead() { assert(false); }
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/clients/Client.h]: Done.

**Comment by eduard-bagdasaryan** [comment]
> We wanted to remove Connection storage to address suspicions that such problems exist, especially since we were adding additional asynchronous calls.

Agreed.
IIRC, our refactoring had 2 steps:
1. Remove 'excessive' checks whether the Connection is not closed (and leave only one such check before, roughly, comm::read()).
2. Remove the Connection object itself from DeferrredRead structures (with the rationale you described above).

File: src/tunnel.cc (hunk 2 above)
Comment (rousskov) [reply src/tunnel.cc:861]: > Your line is the same as mine

GitHub would not highlight the difference if there was none. Your line misspells `Read` as `Dead`.

Comment (eduard-bagdasaryan) [reply src/tunnel.cc]: Sorry, missed it. Done.

File: src/clients/Client.h [hunk 4]
```diff
@@ -155,6 +155,10 @@ class Client:
     size_t calcBufferSpaceToReserve(const size_t space, const size_t wantSpace) const;
 
     void adjustBodyBytesRead(const int64_t delta);
+    /// A callback used by delayRead(); implement in sublclasses calling delayRead().
+    virtual void noteDelayedRead() = 0;
+    /// Defer a noteDelayedRead() call, which must be implemented in subclasses.
```
Comment (rousskov) [inline src/clients/Client.h:160]: The goal here is probably to defer a read; deferring the call is just a (part of the) mechanism doing that. Let's also add a new line, for readability sake.

```suggestion

    /// Defer reading until it is likely to become possible.
    /// Eventually a noteDelayedRead() callback will be called.
```
Please double check that my "likely to become possible" condition is more-or-less accurate. We do not have to be super precise here -- there are probably a lot of variables that control the need for this delay and the timing of the callback -- but we need to give a general idea about what is going on.

File: src/clients/FtpClient.cc [hunk 5]
```diff
@@ -914,6 +916,26 @@ Ftp::Client::maybeReadVirginBody()
 
     initReadBuf();
 
+    delayAwareRead();
+}
+
+void
+Ftp::Client::noteDelayedRead()
```
Comment (rousskov) [inline src/clients/FtpClient.cc:923]: We are notified about an opportunity to do a read, not about a (previously delayed) read that has happened now. I suggest clarifying this by renaming these methods into noteReadChance() or, if we must spell it out, noteDelayAwareReadChance().

**Review by rousskov — CHANGES_REQUESTED** [review]
Please also check the updated PR description and adjust further as needed.

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (up to 004e177).

**Pull request closed without merging**